src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -1389,22 +1389,61 @@ private bool HasIdentityConversionInternal(TypeSymbol type1, TypeSymbol type2)
             return HasIdentityConversionInternal(type1, type2, IncludeNullability);
         }
 
+        /// <summary>
+        /// Returns true if:
+        /// - Either type has no nullability information (oblivious).
+        /// - Both types cannot have different nullability at the same time.
+        /// </summary>
         internal bool HasTopLevelNullabilityIdentityConversion(TypeSymbolWithAnnotations source, TypeSymbolWithAnnotations destination)
         {
             if (!IncludeNullability)
             {
                 return true;
             }
-            return source.IsNullable == destination.IsNullable;
+
+            if (source.NullableAnnotation == NullableAnnotation.Unknown ||
+                destination.NullableAnnotation == NullableAnnotation.Unknown)
+            {
+                return true;
+            }
+
+            if (source.IsPossiblyNullableReferenceTypeTypeParameter() && !destination.IsPossiblyNullableReferenceTypeTypeParameter())
+            {
+                return destination.NullableAnnotation.IsAnyNullable();
+            }
+
+            if (destination.IsPossiblyNullableReferenceTypeTypeParameter() && !source.IsPossiblyNullableReferenceTypeTypeParameter())
+            {
+                return source.NullableAnnotation.IsAnyNullable();
+            }
+
+            return source.NullableAnnotation.IsAnyNullable() == destination.NullableAnnotation.IsAnyNullable();
         }
 
+        /// <summary>
+        /// Returns false if source type can be nullable at the same time when destination type can be not nullable. 
```

Comment:
>Returns false if source type can be nullable at the same time when destination type can be not nullable.  [](start = 12, length = 105)

nit: Same here, the wording could be tweaked instead of simply saying "nullable" as if it's one category.